### PR TITLE
Add utils package

### PR DESCRIPTION
**Pull request by streeve**
Create a separate `utils` package to share capability across Cabana `core` and `cajita`. Currently contains `Version`, `ParameterPack` (this enables #296), and unit test backend files

### CODE REVIEW

File: cajita/src/Cajita_Halo.hpp [hunk 1]
```diff
@@ -14,7 +14,8 @@
 
 #include <Cajita_Array.hpp>
 #include <Cajita_IndexSpace.hpp>
-#include <Cajita_ParameterPack.hpp>
+
+#include <Cabana_ParameterPack.hpp>
```
Comment (sslattery) [inline cajita/src/Cajita_Halo.hpp:18]: Did you mean to include `<Cabana_Utils.hpp>` here?

File: cajita/src/Cajita_Halo.hpp [hunk 2]
```diff
@@ -540,11 +541,12 @@ class Halo
                const Kokkos::View<int* [6], memory_space>& steering,
                const int element_idx,
                const std::integral_constant<std::size_t, 0>,
-               const ParameterPack<ArrayViews...>& array_views )
+               const Cabana::ParameterPack<ArrayViews...>& array_views )
```
Comment (sslattery) [inline cajita/src/Cajita_Halo.hpp:544]: Note that this change of namespace for `ParameterPack` from `Cajita` to `Cabana` is a backwards compatibility break. We are already planning a bunch of small ones anyway elsewhere (e.g. in #342) so this doesn't really bother me that much.

File: utils/unit_test/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,104 @@
+############################################################################
```
Comment (sslattery) [inline utils/unit_test/CMakeLists.txt:1]: We don't need to do it in this PR but we should use this new utils package as an opportunity to unify our test harness and have a single version sitting in `utils/`

File: core/src/Cabana_CommunicationPlan.hpp [hunk 4]
```diff
@@ -12,7 +12,7 @@
 #ifndef CABANA_COMMUNICATIONPLAN_HPP
 #define CABANA_COMMUNICATIONPLAN_HPP
 
-#include <CabanaCore_config.hpp>
+#include <Cabana_config.hpp>
```
Comment (sslattery) [inline core/src/Cabana_CommunicationPlan.hpp:15]: Did you mean to include `Cabana_Utils.hpp`

**Review by sslattery — APPROVED** [review]
Just a few small things. Does this file need to be updated? https://github.com/ECP-copa/Cabana/blob/master/cmake/CabanaConfig.cmakein

File: cajita/src/Cajita_ParameterPack.hpp [hunk 5]
```diff
@@ -12,169 +12,45 @@
...
-    static constexpr std::size_t size = sizeof...( Types );
-};
-
-//---------------------------------------------------------------------------//
 // Static type checker.
-template <class>
-struct is_parameter_pack_impl : public std::false_type
+template <class T>
+struct is_parameter_pack : public Cabana::is_parameter_pack_impl<
+                               typename std::remove_cv<T>::type>::type
```
Comment (dalg24) [inline cajita/src/Cajita_ParameterPack.hpp:22]: Already exists in `utils/src/Cabana_ParameterPack.hpp`

**Comment by dalg24** [comment]
What else than the parameter pack utility do you want to move to that utils package?

Because these changes feel a bit silly if all you need if move parameter pack to Cabana.

File: cajita/src/Cajita_Halo.hpp (hunk 1 above)
Comment (streeve) [reply cajita/src/Cajita_Halo.hpp]: Fixed

File: cajita/src/Cajita_Halo.hpp (hunk 2 above)
Comment (streeve) [reply cajita/src/Cajita_Halo.hpp:544]: Added a wrapper to keep backwards compatibility

File: utils/unit_test/CMakeLists.txt (hunk 3 above)
Comment (streeve) [reply utils/unit_test/CMakeLists.txt:1]: Done

File: core/src/Cabana_CommunicationPlan.hpp (hunk 4 above)
Comment (streeve) [reply core/src/Cabana_CommunicationPlan.hpp]: Fixed

File: cajita/src/Cajita_ParameterPack.hpp (hunk 5 above)
Comment (streeve) [reply cajita/src/Cajita_ParameterPack.hpp:22]: This is already used downstream

**Comment by streeve** [comment]
> What else than the parameter pack utility do you want to move to that utils package?
> 
> Because these changes feel a bit silly if all you need if move parameter pack to Cabana.

Fair point, but we want to use it across both cajita and core. What would you suggest?

And we could do it separately, but it's a good place for the shared testing infrastructure

**Comment by sslattery** [comment]
> > What else than the parameter pack utility do you want to move to that utils package?
> > Because these changes feel a bit silly if all you need if move parameter pack to Cabana.
> 
One thing to note is that I actually use the `ParameterPack` downstream in other libraries for meta-programming to capture variadic parameters on device. I could argue that if I wanted `ParameterPack` and nothing else in Cajita (e.g. I was doing particles only) then I would be building Cajita just for that one class. I can see some of the base tools in `core/src/impl` going in there or tools such as the `Version` code. Also a place to put the shared test harness.

**Comment by dalg24** [comment]
> > What else than the parameter pack utility do you want to move to that utils package?
> > Because these changes feel a bit silly if all you need if move parameter pack to Cabana.
> 
> Fair point, but we want to use it across both cajita and core. What would you suggest?
> 
> And we could do it separately, but it's a good place for the shared testing infrastructure

You already have the Cajita dependency on Cabana, so I'd put it in Cabana.

**Comment by streeve** [comment]
Closing in favor of #363 and #364

**Pull request closed without merging**